modopt/opt/proximity.py

```diff
@@ -688,3 +689,341 @@ def _cost_method(self, *args, **kwargs):
+                    sum_1 = self._compute_theta(data_abs, alpha[last_idx],
+                                                extra_factor).sum()
+                    if (sum_0 <= self.k_value) or (sum_1 >= self.k_value):
+                        found = True
+
+            sum_0 = self._compute_theta(data_abs, alpha[midpoint],
+                                        extra_factor).sum()
+            sum_1 = self._compute_theta(data_abs, alpha[midpoint + 1],
+                                        extra_factor).sum()
+
+            if (sum_0 <= self.k_value) & (sum_1 >= self.k_value):
+                found = True
+
+            elif sum_1 < self.k_value:
+                first_idx = midpoint
+
+            elif sum_0 > self.k_value:
+                last_idx = midpoint
+
+            prev_midpoint = midpoint
+
+        if found:
+            return midpoint, alpha[midpoint], alpha[midpoint + 1], sum_0,\
+                   sum_1
+        else:
+            return -1
+
+    def _find_alpha(self, input_data, extra_factor=1.0):
+        """ Find alpha value to compute theta.
+        This method aim at finding alpha such that sum(theta(alpha)) = k
+        Parameters:
+        -----------
+        input_data: np.ndarray
+            Input data
+        extra_factor: float
+            Potential extra factor for the weights
+        Return:
+        -------
+            alpha: float
+                An interpolation of alpha such that sum(theta(alpha)) = k
+        """
+        data_size = input_data.shape[0]
+
+        # Computes the alpha^i points line 1 in Algorithm 1.
+        alpha = np.zeros((data_size * 2))
+        data_abs = np.abs(input_data)
+        alpha[:data_size] = (self.beta * extra_factor) / \
+                            (data_abs + sys.float_info.epsilon)
+        alpha[data_size:] = (self.beta * extra_factor + 1) / \
+                            (data_abs + sys.float_info.epsilon)
+        alpha = np.sort(np.unique(alpha))
+
+        # Identify points alpha^i and alpha^{i+1} line 2. Algorithm 1
+        _, alpha_0, alpha_1, sum_0, sum_1 = self._binary_search(input_data,
+                                                                alpha,
+                                                                extra_factor)
+
+        # Interpolate alpha^\star such that its sum is equal to k
+        alpha_star = self._interpolate(alpha_0, alpha_1, sum_0, sum_1)
+
+        return alpha_star
+
+    def _op_method(self, data, extra_factor=1.0):
+        """Operator
+
+        This method returns the proximity operator of the squared k-support
+        norm. Implements (Alg. 1) in [M2016].
+
+        Parameters
+        ----------
+        data : np.ndarray
+            Input data array
+        extra_factor : float
+            Additional multiplication factor
+
+        Returns
+        -------
+        np.ndarray proximal map
+
+        """
+        data_shape = data.shape
```

Comment:
Need to add some checks to be sure that the k value is basically not greater than the dimension of the input vector, otherwise setting its maximal value

Reply:
Done